update the encryption spec to use HMAC and hash the entire header

oconnor663 opened the pull request:
I ended up going with the "reserializing the header is allowed approach" rather than framing, but we can go the other way if you prefer. I didn't try to truncate the 32-byte `crypto_auth` output, though that's an option too.

@maxtaco

maxtaco commented on go/saltpack/specs/saltpack_encryption.md:
```diff
@@ -135,8 +139,31 @@ the same, clients may indicate that a message is "intentionally anonymous" as
+  this.)
+- Positive integers must used the unsigned integer formats.
+
+The **MAC key** is a 32-byte secret key, unique to each recipient. It's
```
say what the nonce here is.

maxtaco commented on go/saltpack/specs/saltpack_encryption.md:
```diff
@@ -160,72 +187,79 @@ recipients, then she will have the key, and she could modify the payload. The
-- Encrypt that hash in a NaCl box, using the sender and recipient's long-term
-  keys. See [Nonces](#nonces) below.
-- The authenticator is the first 16 bytes of that box.
+- Serialize the packet number as a 64-bit unsigned big-endian integer, where
```
Just use the payload box nonce here, so 16 zeros before the 64-bit unsigned big-endian integer.

maxtaco commented on go/saltpack/specs/saltpack_encryption.md:
````diff
@@ -148,7 +175,7 @@ A payload packet is a MessagePack list with these contents:
 ```
 
-- The **authenticators list** contains 16-byte Poly1305 tags, one for each
+- The **authenticators list** contains 32-byte HMAC tags, one for each
````
they also auth the header

maxtaco commented on go/saltpack/specs/saltpack_encryption.md:
```diff
@@ -160,72 +187,79 @@ recipients, then she will have the key, and she could modify the payload. The
-list** uses 1, the next uses 2, and so on. For each **payload secretbox**, the
-nonce is the same as for the associated authenticators. The strict ordering of
-nonces should make it impossible to drop or reorder any payload packets.
+The 24-byte nonce for each public-key box is then concatenation of `P` and an
```
Can you call it *P* rather than `P`?

oconnor663 commented:
TODO: add a link to the msgpack spec

maxtaco commented:
Other TODOS:
- Make a forward reference the first time you mention **payload key** since it's out of the blue.

maxtaco commented on go/saltpack/specs/saltpack_encryption.md:
```diff
@@ -160,72 +187,79 @@ recipients, then she will have the key, and she could modify the payload. The
 recipients from each other.
 
-We compute each recipient's authenticator in three steps:
+We compute the authenticators in three steps:
```
Right above here, you say "protecting the recipient from each other." That used to be true, but we're doing double-duty here and also protecting the header. And this is the only we we protect the header (only the payload which is double-protected).

maxtaco commented:
Another high-level comment is that the `## Header Packet` section seems to ramble a little bit now.  I wonder if it's possible to add more structure there?

oconnor663 commented:
I split the Header section up a bit, and added explicit instructions for the sender and recipient. Maybe easier to read at https://github.com/keybase/client/blob/jack/test/go/saltpack/specs/saltpack_encryption.md.

maxtaco commented on go/saltpack/specs/saltpack.md:
```diff
@@ -18,7 +18,7 @@ We start with some simplifying decisions:
   formatting.
 
-Thus, "SaltPack" (pronounced "NaCl pack"). The spec is in three parts:
+Thus, "saltpack" (pronounced "NaCl pack"). The spec is in three parts:
```
nuke joke, sorry!

maxtaco commented on go/saltpack/specs/saltpack_encryption.md:
```diff
@@ -113,30 +120,81 @@ A recipient pair is a two-element list:
+   [`crypto_box_keypair`](http://nacl.cr.yp.to/box.html).
+3. Encrypt the sender's long-term public key using
+   [`crypto_secretbox`](http://nacl.cr.yp.to/secretbox.html) with the **payload
+   key**, to create the **sender secretbox**. (See [Nonces](#nonces).)
```
Instead say `(See [Nonces](#nonces) below for a description of which nonce to used)`.

oconnor663 replied:
Every time, or just the first time?

maxtaco replied:
just the first. Sorry for the exceptional first case :)

oconnor663 replied:
IS THERE NO END TO YOUR MADNESS

maxtaco commented on go/saltpack/specs/saltpack_encryption.md:
```diff
@@ -113,30 +120,81 @@ A recipient pair is a two-element list:
+   [`crypto_box_beforenm`](http://nacl.cr.yp.to/box.html) with the **ephemeral
+   public key** and the sender's private key.
+7. Try to open each of the **payload key boxes** in the recipients list using
+   [`crypto_box_afternm`] and the shared secret from #6. (See
```
URL for this one?

oconnor663 replied:
Woops!

maxtaco commented on go/saltpack/specs/saltpack_encryption.md:
```diff
@@ -113,30 +120,81 @@ A recipient pair is a two-element list:
+anonymous to Mallory. If Alice wants to be anonymous to Bob as well, she can
+reuse the ephemeral key as her sender key. When the ephemeral key and the
+sender key are the same, clients may indicate that a message is "intentionally
+anonymous" as opposed to "from an unknown sender".
```
I like these new lists, they're easier to follow.

maxtaco commented on go/saltpack/specs/saltpack_encryption.md:
```diff
@@ -113,30 +120,81 @@ A recipient pair is a two-element list:
+4. For each recipient, encrypt the **payload key** using
+   [`crypto_box`](http://nacl.cr.yp.to/box.html) with the recipient's public
+   key and the ephemeral private key. (See [Nonces](#nonces).) Assemble these
+   into the **recipients list**.
```
Say to optionally include the receiver KID with each ciphertext?

maxtaco commented on go/saltpack/specs/saltpack_encryption.md:
```diff
@@ -113,30 +120,81 @@ A recipient pair is a two-element list:
+6. Precompute the ephemeral shared secret using
+   [`crypto_box_beforenm`](http://nacl.cr.yp.to/box.html) with the **ephemeral
+   public key** and the sender's private key.
+7. Try to open each of the **payload key boxes** in the recipients list using
```
As before, do you need to iterate if the your KID is explicitly listed?

oconnor663 replied:
This one's clarified in the paragraph just below.

maxtaco commented:
made a few more comments on the new stuff....

oconnor663 commented:
Landed f2f2281a37083e238f50f462d30219dcc02a80a1.

oconnor663 commented:
Oops I mean 316ddaf5329bb73646f5d2df723c86d3705ae020. I got my branches confused.
